### ISSUE TITLE
[Bug] Type definitions do not specify `Promise<T>` as possible return value

### ISSUE THREAD

**Issue by ghengeveld** [storybookjs/testing-library#10]
Wrapped methods can return a Promise, but the type definitions do not reflect that. This causes TS to complain when trying to use e.g. `await userEvent.click(...)`:

```
'await' has no effect on the type of this expression.
```

**Comment by yannbf** [comment]
Unfortunately there is no way to do declaration merging/module augmentation in the types from @testing-library/dom because they expose types and not interfaces. I think the only solution would be to copy paste the original types and change them. That brings trouble when thinking of upgrading testing lib dependency, but at least we'd be in control. Not sure if I like this direction though! @ghengeveld @shilman

**Comment by bilalq** [comment]
Can this be managed by improving the type of the [`instrument` function from `@storybook/instrumenter`](https://github.com/storybookjs/storybook/blob/next/lib/instrumenter/src/instrumenter.ts#L541)?

Today, the signature of the instrument function is:

```typescript
export function instrument<TObj extends Record<string, any>>(
  obj: TObj,
  options: Options = {}
): TObj
```

It's declaring that it returns an object of the same type as its input, but that's just not the case. I didn't spend too long digging through the code, but it seems like it recursively transforms each function to return a promise in the given object.

This can be achieved by leveraging recursive conditional types in TypeScript. Here's an example showcasing this:

```typescript
/**
 * A utility type that transforms a function type into one that returns a
 * Promise type of the existing value type.
 */
type PromisifiedReturn<T extends (...args: any) => any> =
  ReturnType<T> extends PromiseLike<unknown>
    ? T
    : (...a: Parameters<T>) => Promise<ReturnType<T>>

/**
 * A recursive utility type that transforms all functions types nested within
 * the object to return Promises instead of their existing value types.
 */
type RecursivelyTransformedObject<T> = {
  [K in keyof T]: 
    // If the value is a function, we transform its type to be one that returns a Promise
    T[K] extends (...args: any) => any 
    ? PromisifiedReturn<T[K]>

    // If it's an Array, we recursively transform its elements
    : T[K] extends (infer U)[]
    ? RecursivelyTransformedObject<U>[]

    // If it's an Object, we recursively transform its properties
    : T[K] extends object
    ? RecursivelyTransformedObject<T[K]>

    // Otherwise, we just keep the type as it is
    : T[K]
}

// Just some filler functions so we can have an interesting input object later
const click = (element: Element, _init?: MouseEventInit | undefined): void =>
  console.log(element)
const dataReturningFunction = (num: number) => num * 2
const asyncVoidFunction = (num: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, num))
const asyncDataReturningFunction = (str: string): Promise<boolean> =>
  Promise.resolve(str.startsWith('/'))

// Here's a fairly complex object to instrument
const inputObj = {
  booleanThatIsNotAFunction: true,
  numThatIsNotAFunction: 100,
  click,
  inlineFn: () => 100 as const,
  asyncVoidFunction,
  nestedObj: {
    extraNestedList: [
      {
        anotherFn: () => 10,
        anotherAsyncFn: () => Promise.resolve(false),
      },
      () => 'one more fn',
      'string that is not a function'
    ],
    dataReturningFunction,
    asyncDataReturningFunction,
  },
} as const

const instrument = <T extends object>(
  _obj: T
): RecursivelyTransformedObject<T> => {
  throw new Error('Not implementing this here')
}

const instrumentedObj = instrument(inputObj)

const inAsyncUsage = async () => {
  await instrumentedObj.click(document.querySelectorAll('*')[0], {}) // Return type is now Promise<void>
  await instrumentedObj.asyncVoidFunction(10) // Return type remains Promise<void>
  await instrumentedObj.nestedObj.extraNestedList[0].anotherFn() // Return type is now Promise<number>
  await instrumentedObj.nestedObj.extraNestedList[1]() // Return type is now Promise<string>
  instrumentedObj.booleanThatIsNotAFunction // This boolean remains a Boolean
  instrumentedObj.numThatIsNotAFunction.toFixed(0) // This number remains a number
  instrumentedObj.nestedObj.extraNestedList[2].toUpperCase() // This string remains a string
}
```

You can experiment with this in the TypeScript playground using [this link here](https://www.typescriptlang.org/play?#code/PQKhCgAIUhBSCuAXAlgGxUgnpbAHAU1wAsBDJXAJ1IDsBnAMwHtKBbOyUyBhGgY1RMauLIUgoaSJpCFEkZCpQJIElepygwACpSasUdOaKJMGJIgQAeB1DQDmkAG6k0CI4QB0m4OHxEdegYoDCgEACYASsqqNAA8ACqQVkgENGEcABQe2aSUdnQAXJw0WACUkAC8AHzFWDUVUJBRKmrxxgk1yanpkAH6hgAyKADWBLG8wzRMAO40VY2QkAD8kPELkEVZOUVauaSsygSUdB3l1b26-WPNMW2EHfPgoBDQcJBKfKp0KI5EyOiYHB+EjkKi0RgsdicNBobi8AQoIQcPwcGgEOgpMKQaaYYgSTTmGQAIwAVgQBLhpEoWsI+gZ0eJ6ClSFjTOYUJQktYMRIHM5XO50V5oD5gVFPscfgQ0Fh4tR6Mw2OEAPKk8lIDqVSAAb0aAG0ANKMyCjLBs+IAXSK62AwEgAEkzPIiPy3OIOFwePxBDQADTYuTyiFscRIZHGSmQIkmNEgxTRNQei6BQzreKGi1clJpTLZDy5fJFWhlSo1YuQdYrOnfELhG5qBIZ+aLRa2h1mTAAcg9wlglGoWH90yIHy+UplYIVkNDHGlBAOkjo6yK6YNma6OcgGQkDCOkAAqqU9RbK01yWPfjK5eDFQcwqqyQJYvuqsebXbHaHu8VIA-1UOR3PSVLyBINbxnSA8F0QhKFQdFl1WDMs26DgmDVART3FC9pVlMDIRVdCNVXC1mxbNtlWdSgcUMADIBJBAMRNAgCDwQlgVIDhMHdBDiPAABfcAnjtAApBiKDoPQiBCGE9y9BEkUgCSA0gPhaEgMhfh-CQUiUHl7EZPBkGJR8KDQcgjnAPgkQoPgMD4YYtQyOcFyQIoAFE0HnVIkH9AB9CRMCWIoAFkmAYgg3N+SR7RoLiAB9EDSAgQjRMJSiKRwmBQLFqkaKz6CYTyPDQJg7CczyXNKSzrMgMJyFIetYvsAAxeEfUcmgEFYIpOtYaNKDOGpeteAAmaqmU4OgsH4AA1LKwla71EWECot16nquv69LkyuWJMuy+pSLRaYdvpDIMl0wrfkGxTlHiFADjCpALvRK6CH9XrSiq-LGI46a+AAEXqxreUW+SVq3DFKCKKHeW26sxiJJhCoIWhDsaBGPEutBfgyKGPAxXIwwAdVxDJO2ATsvsEn6KAkQykAfLVdUWJGUdoeIFHtOgADkmCQWAwZ9IokEoNxfUaXrOfIbm+YFoXlqKABGAAGFWJcWWyUHsjXGQwNFmpoTYbtVlXJpU6zdb+2b5oVoRdbRDECJJIoWZbZJqB59FMSGDEij1dZFjdlsW1ofniCOQ3jdLSBVd1kPFjDyjYCm-go63G7Mex3GGBcQxSnjls+MLxYMhuztZEgVgWCkmhOxLyBu1F3k43dSApgoT02uWzt1gtQu6qQBqEyauw7b9dZrcB4GR9B7v7caYv+PN2maZqiQoa67ytQSZDNzQkyqgyRpfIPldwG2rDgJw68pyVe9CM1c43fkXQTuOyA3P7Fhybl8RWDwBVbyLd5AGHUkcAgVN+JrwmhvUWW9JDOy1HAsWLltw0AZg+b668aAp3+vuOgpA7BEFWlPDOMc3akGmKQLiKCEGYgfB4LW9kMhhCYJ8FyHgACObhKBYAAMpzgECwWAMJyYgCpnqFW-cdR8XKG2RqIgxBgKmCdBGe15qkSoTQumTJUHeWdvmVOfA5rZXHhkVW8i7SKOBEoVgND1DqP2mELR1DaF6PoYYx2DDSQeA9qQL2Tswi+yQFIi0+YO4R0oIbMukAFEjyUUQFRMxTqGFiL1fqridGMk3i5Lx3tDH+MCT7GweolYWlifEmkiS26qNSWMWG9hSJ0LyQ-EkHg2aeQ5lzXm-NBbz2EG2TmYDOmo2EHYhxSYABCyMuk0EaC0gxbSPBSx6XLfpS0hAeCkM1FAlhwgZBVlY1YeJUSbT3BMjenB27nMoAsjxrTGHeMKZYUWASCnBNKSNcJUh9x4BggAYQ4gQSpdphkcEaQ4S56guCQugUAA) and hovering over the different properties to see the types the compiler infers them to be.

Note that the conditional nature of patching the return types can also be managed by discriminating function signatures based on the `intercept` boolean.

**Comment by donaldpipowitch** [comment]
Would be lovely to fix the types. Another use case are users of [`@typescript-eslint/await-thenable`](https://typescript-eslint.io/rules/await-thenable/).

**Comment by kasperpeulen** [comment]
@shilman @vanessayuenn  
I think this would be required for the first-class-typescript project.
In my QA project I see a lot of not await-ed expect, causing the interaction debugger to be buggy.

Tools (for example ESLint, default editor analytics in webstorm) can warn for any not awaited promise. 
So if we have this, people are more likely to write play functions that are not buggy when used with the interactions debugger.

**Comment by fabioimpe** [comment]
+1 on this.
Are there any workaround in the meanwhile?

**Comment by smagch** [comment]
I'm using the following approach:

```tsx
await Promise.resolve(userEvent.click(button));
await Promise.resolve(expect(children).not.toBeVisible());
```

It's not sophisticated, but serves well as a temporary solution.

**Comment by psychobolt** [comment]
My solution, if typescript-eslint warns...

> I'm using the following approach:
> 
> ```tsx
> await Promise.resolve(userEvent.click(button));
> await Promise.resolve(expect(children).not.toBeVisible());
> ```
> 
> It's not sophisticated, but serves well as a temporary solution.

```
await Promise.resolve(userEvent.click(button)); // eslint error
```

```
Returning a void expression from an arrow function shorthand is forbidden. Please add braces to the arrow function.eslint[@typescript-eslint/no-confusing-void-expression](https://typescript-eslint.io/rules/no-confusing-void-expression)
```

I had to configure the rule to allow arrow functions:

```json
"@typescript-eslint/no-confusing-void-expression": [
  "error",
  {
    "no-confusing-void-expression": true
  }
],
```
```js
await Promise.resolve(() => userEvent.click(button)); // no eslint error!
```

**Comment by yannbf** [comment]
Hey there!
[@storybook/testing-library v0.2.0](https://github.com/storybookjs/testing-library/releases/tag/v0.2.0) has been released with upgraded versions of user-event and testing-library/dom. This means that `userEvent` will now have correct types with everything being a promise. Please give it a try and let us know if you encounter any problems!

**Comment by lewisflude** [comment]
> Hey there! [@storybook/testing-library v0.2.0](https://github.com/storybookjs/testing-library/releases/tag/v0.2.0) has been released with upgraded versions of user-event and testing-library/dom. This means that `userEvent` will now have correct types with everything being a promise. Unfortunately this only applies for `userEvent` and not for `expect`.
> 
> Please give it a try and let us know if you encounter any problems!

It's working great for useEvent and not for expect, exactly as you describe! Thanks :)

**Comment by ubugnu** [comment]
for `expect` the solution is to use `.resolves` and to wrap the expected value in a Promise:

```js
test('resolves to lemon', async () => {
  await expect(Promise.resolve('lemon')).resolves.toBe('lemon');
  await expect(Promise.resolve('lemon')).resolves.not.toBe('octopus');
});
```
Source: https://jestjs.io/docs/expect#resolves

**Comment by yannbf** [comment]
Hey everyone, I'm closing this issue as it's been resolved in [@storybook/testing-library v0.2.0](https://github.com/storybookjs/testing-library/releases/tag/v0.2.0). The issue with `expect` relates to another library, which you can find and follow its progress here:
https://github.com/storybookjs/jest/issues/32
